Snapshot obfuscate errors out + optimizing obfuscate

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
Errors out as follows:
  File "/opt/ptl/bin/pbs_snapshot", line 188, in obfuscate_snapshot_wrapper
    obj.obfuscate_snapshot(snap_name, map_file, sudo_val)
  File "/opt/ptl/lib/python3.6/site-packages/ptl/utils/pbs_snaputils.py", line 508, in obfuscate_snapshot
    self.obfuscate_acct_logs(snap_dir, sudo_val)
  File "/opt/ptl/lib/python3.6/site-packages/ptl/utils/pbs_snaputils.py", line 430, in obfuscate_acct_logs
    self._obfuscate_acct_file(attrs_to_obf, acct_fpath)
  File "/opt/ptl/lib/python3.6/site-packages/ptl/utils/pbs_snaputils.py", line 357, in _obfuscate_acct_file
    with open(file_path, "r") as fd, open(fout, "w") as fdout:
FileNotFoundError: [Errno 2] No such file or directory: 'new_snapshot_infra_bigmem/server_priv/accounting/new_snapshot_infra_bigmem/server_priv/accounting/20170617'

- Also, --obfuscate had a performance inefficiency: towards the end, it would open, parse, modify and write all of the files in the snapshot multiple times, once for each key-val pair in the obfuscation map, in order to obfuscate each value. I changed it to open the file, swap all the values in the map and write back, just once.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
DshUtils.listdir returns full paths, not just file names, which is what we were assuming in the obfuscation codepath.

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
[testsnap.log](https://github.com/openpbs/openpbs/files/6052817/testsnap.log)

[test_capture_server.log](https://github.com/openpbs/openpbs/files/6126788/test_capture_server.log)


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

npadole20 commented on test/tests/functional/pbs_snapshot_unittest.py:
```diff
@@ -257,7 +257,15 @@ def test_capture_server(self):
+                    except AssertionError:
+                        # Check if this was a core file, which would
+                        # explain why it was captured
+                        if CORE_DIR in file_fullpath:
```
As part of capture_server(), will we also capture core files in another directory apart from the server ones?

agrawalravi90 replied:
No we don't, we only capture server's core files. The thing is, we store them in a separate path inside the snapshot: <snapshot name>/core_file_bt
The test is checking that the only data captured was inside directories that shared a common path with <snapshot name>/server, which 'core_file_bt' will not. So, that's why I made the exception.

Did you want me to check that CORE_DIR didn't capture core files from other daemons?

npadole20 replied:
Yes. I think we should do that. It should be simple as we are creating a different directory for every daemon. If we have only the server_priv directory in core_file_bt/ that should be good enough check.

agrawalravi90 commented:
@npadole20 thanks for your review, I've addressed your comment, attached test log after making the latest changes, i had a core file in mom_priv as well as server_priv and it was able to capture only server .. please provide further feedback

npadole20 reviewed (approved):
Thanks for making the changes.

agrawalravi90 commented:
@openpbs/openpbs-maintainers  need a maintainer review for this
@npadole20 thanks for the review!

agrawalravi90 commented:
@openpbs/openpbs-maintainers gentle reminder, thanks
